### ISSUE TITLE
Ansible errors

### ISSUE THREAD

**Issue by prasadtalasila** [AutolabJS/AutolabJS#57]
The following errors are known to occur during install/uninstall process.

1. Only one execution node gets installed. We need multiple execution nodes in place. We can create five execution nodes by default.    
1. During uninstall the keys of the main_server (JavaAutolab/deploy/keys/main_server/id_rsa*) are not removed. This is leading to an error during reinstall process.    
1. The install script needs to consider the possibility of rerun and skip steps that have already been completed. We need to use the conditional execution feature of Ansible ([include...when](http://docs.ansible.com/ansible/latest/include_module.html): clause of a task) in order to have this feature in place.

**Comment by prasadtalasila** [comment]
Use playbook blocks feature to solve problem-3.

**Comment by prasadtalasila** [comment]
[deploy/keys](https://github.com/prasadtalasila/JavaAutolab/tree/dev/deploy/keys) are used to hold the keys of main_server and load_balancer for Gitlab. These keys enable main_server and load_balancer to SSH login to GitLab and download the necessary code base for backup. The load balancer uses such a facility to backup the student commits that received best score. The main_server is supposed use this facility in revaluation.

At present, only load_balancer is utilising its Gitlab key properly. The revaluation functionality of main_server is broken. Only after that facility is fixed, can we utilise the SSH login facility available to main_server.

On a crucial note, the execution nodes also need SSH login facility to Gitlab. The keys are stored in /root/.autolab/keys. It is better to generate the keys and configs locally and keep them in deploy/keys/execution_nodes and deploy/configs/execution_nodes.

We need to make provision for creating multiple execution nodes on demand. As new execution nodes are configured, we need to create the necessary keys and configs for them in deploy/ directory.

**Comment by prasadtalasila** [comment]
At present, the following keys are not updated for each install
```shell
main_server/ssl
load_balancer/ssl
execution_nodes/ssl
```

Ideally a fresh set of keys are generated for each new install before ansible playbook is called. The key generation functionality should be taken out of Ansible into a dedicated script.

A suggested directory structure for keys is as follows.
```shell
deploy/keys/
       gitlab/
              main_server/
              load_balancer/
              execution_nodes/
       main_server/
       load_balancer/
       execution_nodes/
```
The `deploy/keys/gitlab` directory stores the key pairs used by all the Autolab components for logging into Gitlab.
The `deploy/<component_name>` directory contains the SSL keys used by the express servers of each of the components. These component keys have to be volume mapped to docker containers of the Autolab components at `/etc/<component_name>` locations.

**Comment by prasadtalasila** [comment]
lines - 218 to 228 of playbook.yml install docker software. This is required only when the execution nodes are to be run on Machine-2. Thus playbook.yml assumes a two machine deployment model. These lines are not needed for single machine deployment.

**Comment by prasadtalasila** [comment]
The SSL keys inserted into Gitlab are all called **rootkey**. Appropriate naming of keys would be good. The suggested names are:
main_server --> for root key of main server in Gitlab
load_balancer --> for root key of load balancer in Gitlab
execution_node-ip-port --> for root key of one instance of execution node

**Comment by prasadtalasila** [comment]
playbook*.yml requires the following directories to be empty.
```
deploy/keys/main_server
deploy/keys/load_balancer
```
Usually after one installation, two files, namely id_rsa and id_rsa.pub exist in those two directories. If we forget to delete these key files before reinstallation attempt, the reinstallation fails.

For now, the remedy is to execute the following commands before any reinstallation attempt.
```
rm deploy/keys/main_server/id_rsa*
rm deploy/keys/load_balancer/id_rsa*
```

**Comment by prasadtalasila** [comment]
Use the technique shown [here](http://stackoverflow.com/questions/30494050/how-to-pass-environment-variables-to-docker-containers) to pass container environment variables through a config file.

**Comment by prasadtalasila** [comment]
With the code in a301bfd branch, the ansible script fails with the following message log.

```shell
PLAY [executionnodes] **************************************************************************************************

TASK [Gathering Facts] **************************************************************************************************
ok: [10.1.9.207]

TASK [Get Docker installation script] **************************************************************************************************
ok: [10.1.9.207]

TASK [Install Docker] **************************************************************************************************
changed: [10.1.9.207]

TASK [Start docker] **************************************************************************************************
ok: [10.1.9.207]

TASK [mkdir for ssh keys] **************************************************************************************************
 [WARNING]: Consider using file module with state=directory rather than running mkdir

changed: [10.1.9.207]

TASK [Copy SSH Key for GitLab] **************************************************************************************************
fatal: [10.1.9.207]: FAILED! => {"changed": true, "cmd": "cp ./keys/gitlab/execution_nodes/execution_node_1/id_rsa* /root/.autolab/keys/execution-node-10.1.9.207-8082/", "delta": "0:00:00.003423", "end": "2017-08-18 18:17:57.714047", "failed": true, "rc": 1, "start": "2017-08-18 18:17:57.710624", "stderr": "cp: cannot stat './keys/gitlab/execution_nodes/execution_node_1/id_rsa*': No such file or directory", "stderr_lines": ["cp: cannot stat './keys/gitlab/execution_nodes/execution_node_1/id_rsa*': No such file or directory"], "stdout": "", "stdout_lines": []}
	to retry, use: --limit @/home/autolabjs/AutolabJS/deploy/playbook.retry

PLAY RECAP **************************************************************************************************
10.1.9.207                 : ok=11   changed=4    unreachable=0    failed=1   

```

**Comment by prasadtalasila** [comment]
adding ssl keys to gitlab uses deprecated coding idiom. The following message is shown for all ssl key addition cases.
```shell
[DEPRECATION WARNING]: Supplying headers via HEADER_* is deprecated and will be removed in a future version. Please use `headers` to supply 
headers for the request.
This feature will be removed in a future release. Deprecation warnings can be disabled by setting 
deprecation_warnings=False in ansible.cfg.
ok: [localhost]
```

**Comment by prasadtalasila** [comment]
Building docker containers takes a long time. It is better to parallelize the container building work to reduce the installation time. Ansible [`async`](http://docs.ansible.com/ansible/latest/playbooks_async.html) and `poll` commands are useful for our task.

**Comment by prasadtalasila** [comment]
Most of the above mentioned problems have been fixed. The pending ones have been moved to [feature requests](https://github.com/AutolabJS/AutolabJS/wiki/Feature-Requests) page.
